File: jenkins-plugin/src/test/java/org/eazyportal/plugin/release/jenkins/action/SetReleaseVersionActionFactoryTest.java
package org.eazyportal.plugin.release.jenkins.action;

import org.eazyportal.plugin.release.core.action.SetReleaseVersionAction;
import org.eazyportal.plugin.release.core.scm.ConventionalCommitType;
import org.eazyportal.plugin.release.core.scm.ScmActions;
import org.eazyportal.plugin.release.core.scm.model.ScmConfig;
import org.eazyportal.plugin.release.core.version.ReleaseVersionProvider;
import org.eazyportal.plugin.release.core.version.VersionIncrementProvider;
import org.eazyportal.plugin.release.jenkins.ReleaseStepConfig;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

class SetReleaseVersionActionFactoryTest {

    @Mock
    private transient ReleaseStepConfig releaseStepConfig;
    @Mock
    private transient ReleaseVersionProvider releaseVersionProvider;
    @Mock
    private transient VersionIncrementProvider versionIncrementProvider;

    @InjectMocks
    private SetReleaseVersionActionFactory underTest;

    @BeforeEach
    void setUp() {
        MockitoAnnotations.openMocks(this);
    }

    @Test
    void test_create() {
        // GIVEN
        ScmActions scmActions = mock(ScmActions.class);

        // WHEN
        when(releaseStepConfig.getConventionalCommitTypes()).thenReturn(ConventionalCommitType.getDEFAULT_TYPES());
        when(releaseStepConfig.getScmConfig()).thenReturn(ScmConfig.getGIT_FLOW());

        // THEN
        SetReleaseVersionAction actual = underTest.create(scmActions);

        assertThat(actual).hasNoNullFieldsOrProperties();

        verifyNoInteractions(scmActions, releaseVersionProvider, versionIncrementProvider);
        verify(releaseStepConfig).getConventionalCommitTypes();
        verify(releaseStepConfig).getScmConfig();
        verifyNoMoreInteractions(releaseStepConfig);
    }

}

File: gradle-plugin/src/test/kotlin/org/eazyportal/plugin/release/gradle/action/SetReleaseVersionActionFactoryTest.kt
package org.eazyportal.plugin.release.gradle.action

import org.assertj.core.api.Assertions
import org.eazyportal.plugin.release.core.project.ProjectActionsFactory
import org.eazyportal.plugin.release.core.scm.ConventionalCommitType
import org.eazyportal.plugin.release.core.scm.ScmActions
import org.eazyportal.plugin.release.core.scm.model.ScmConfig
import org.eazyportal.plugin.release.core.version.model.VersionIncrement.MAJOR
import org.eazyportal.plugin.release.gradle.model.EazyReleasePluginExtension
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

internal class SetReleaseVersionActionFactoryTest {

    private companion object {
        @JvmStatic
        fun create() = listOf(
            Arguments.of(listOf<ConventionalCommitType>()),
            Arguments.of(listOf(ConventionalCommitType(listOf("dummy"), MAJOR))),
            Arguments.of(ConventionalCommitType.DEFAULT_TYPES)
        )
    }

    private lateinit var underTest: SetReleaseVersionActionFactory

    @BeforeEach
    fun setUp() {
        underTest = SetReleaseVersionActionFactory()
    }

    @MethodSource("create")
    @ParameterizedTest
    fun test_create(conventionalCommitTypes: List<ConventionalCommitType>) {
        // GIVEN
        val extension = mock<EazyReleasePluginExtension>()

        val projectActionsFactory = mock<ProjectActionsFactory>()
        val scmActions = mock<ScmActions>()
        val scmConfig = mock<ScmConfig>()

        // WHEN
        whenever(extension.conventionalCommitTypes).thenReturn(conventionalCommitTypes)
        whenever(extension.projectActionsFactory).thenReturn(projectActionsFactory)
        whenever(extension.scmActions).thenReturn(scmActions)
        whenever(extension.scmConfig).thenReturn(scmConfig)

        // THEN
        val actual = underTest.create(extension)

        Assertions.assertThat(actual).hasNoNullFieldsOrProperties()
    }

}